File: Week 3/Python Codes/DetectLoopLinkedList.py
# Node Class
class Node:
    def __init__(self, data):   # data -> value stored in node
        self.data = data
        self.next = None

def detectLoop(head):
    slow_pointer = head
    fast_pointer = head
    while (slow_pointer and fast_pointer and fast_pointer.next):
        slow_pointer = slow_pointer.next
        fast_pointer = fast_pointer.next.next
        if slow_pointer == fast_pointer:
            return True
    return False

File: Week 3/C++ Codes/DetectLoopLinkedList.cpp
#include<bits/stdc++.h>
using namespace std;

struct Node
{
    int data;
    struct Node *next;
    Node(int x) {
        data = x;
        next = NULL;
    }
};

bool detectLoop(Node* head)
{
    struct Node* slowPointer = head;
    struct Node* fastPointer = head;
    while (slowPointer != NULL and fastPointer != NULL and fastPointer->next != NULL) {
        slowPointer = slowPointer->next;
        fastPointer = fastPointer->next->next;
        if (slowPointer == fastPointer)
            return true;
    }
    return false;
}
